azurelinuxagent/ga/collect_logs.py

```diff
@@ -324,30 +331,57 @@ def daemon(self):
+
         return metrics
 
     def _send_telemetry(self, metrics):
         for metric in metrics:
             report_metric(metric.category, metric.counter, metric.instance, metric.value, log_event=self.__log_metrics)
 
     def _verify_memory_limit(self, metrics):
-        current_usage = 0
+        current_anon_and_swap_usage = 0
+        current_cache_usage = 0
+        memory_throttled_events = 0
         for metric in metrics:
-            if metric.counter == MetricsCounter.TOTAL_MEM_USAGE:
-                current_usage += metric.value
+            if metric.counter == MetricsCounter.ANON_MEM_USAGE:
+                current_anon_and_swap_usage += metric.value
             elif metric.counter == MetricsCounter.SWAP_MEM_USAGE:
-                current_usage += metric.value
-
-        if current_usage > LOGCOLLECTOR_MEMORY_LIMIT:
-            msg = "Log collector memory limit {0} bytes exceeded. The max reported usage is {1} bytes.".format(LOGCOLLECTOR_MEMORY_LIMIT, current_usage)
+                current_anon_and_swap_usage += metric.value
+            elif metric.counter == MetricsCounter.CACHE_MEM_USAGE:
+                current_cache_usage = metric.value
+            elif metric.counter == MetricsCounter.MEM_THROTTLED:
+                memory_throttled_events = metric.value
+
+        mem_limit_exceeded = False
+        if current_anon_and_swap_usage > LOGCOLLECTOR_ANON_MEMORY_LIMIT:
+            mem_limit_exceeded = True
+            msg = "Log collector anon + swap memory limit {0} bytes exceeded. The reported usage is {1} bytes.".format(LOGCOLLECTOR_ANON_MEMORY_LIMIT, current_anon_and_swap_usage)
             logger.info(msg)
-            add_event(
-                name=AGENT_NAME,
-                version=CURRENT_VERSION,
-                op=WALAEventOperation.LogCollection,
-                message=msg)
+            add_event(name=AGENT_NAME, version=CURRENT_VERSION, op=WALAEventOperation.LogCollection, message=msg)
+        if current_cache_usage > LOGCOLLECTOR_CACHE_MEMORY_LIMIT:
+            mem_limit_exceeded = True
+            msg = "Log collector cache memory limit {0} bytes exceeded. The reported usage is {1} bytes.".format(LOGCOLLECTOR_CACHE_MEMORY_LIMIT, current_cache_usage)
+            logger.info(msg)
+            add_event(name=AGENT_NAME, version=CURRENT_VERSION, op=WALAEventOperation.LogCollection, message=msg)
+        if memory_throttled_events > LOGCOLLECTOR_MAX_THROTTLED_EVENTS:
+            mem_limit_exceeded = True
+            msg = "Log collector memory throttled events limit {0} exceeded. The reported number of throttled events is {1}.".format(LOGCOLLECTOR_MAX_THROTTLED_EVENTS, memory_throttled_events)
+            logger.info(msg)
+            add_event(name=AGENT_NAME, version=CURRENT_VERSION, op=WALAEventOperation.LogCollection, message=msg)
+
+        if mem_limit_exceeded:
             os._exit(GRACEFUL_KILL_ERRCODE)
```

Comment:
Created this task to move the graceful kill logic to the log collector thread itself, so that we can exit cleanly.
https://dev.azure.com/msazure/One/_workitems/edit/29146805